### PR TITLE
Blog post: Conditional Tracer with APM on OpenShift

### PR DESCRIPTION
**Pull request by jpkrohling**

### CODE REVIEW

**Comment by jpkrohling** [comment]
@objectiser , @pavolloffay , could you please review this?

@theute , @josejulio , @jkandasa , would any of you also review *and* try the steps from the blog post?

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+== Wrapping up
+
+Deciding whether a transaction should be traced or not is a decision that should
+be outsourced as much as possible to the underlying Tracer, possibly with data
+coming from different sources. Having a simple switch on the code loading the
+Tracer, however, might prove invaluable to admins in determined situations.
+Given the simplicity of this solution, there's no reason not to have it!
+
+The full source code for this example can be found at
+link:https://github.com/hawkular/conditional-tracer[github.com/hawkular/conditional-tracer].
```
Comment (jpkrohling) [inline src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: Either this repository needs to be created, or I need to change the blog post to point to the repository at under my personal account: https://github.com/jpkrohling/conditional-tracer

**Comment by jpkrohling** [comment]
cc @jstrachan , as promised, here's the blog post in case you want to review it as well.

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+the percentage in steps. Similarly, if we are experiencing traffic peaks and need
+every resource available, we might enable tracing only to a small percentage of
+the incoming requests.
+
+For the simple scenario, let's call it "percentage based sampling", we have a few
+alternatives on how to implement it:
+
+=== Application-based decision
+
+On this scenario, we embed the logic into the application, using the percentage
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: "In this scenario"

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+scenarios.
+
+Another solution is to externalize this into a Configuration Map (ConfigMap),
+but then, we also need to let our application listen to events related to this
+ConfigMap. For a microservice, we might end up with a bunch of unrelated
+dependencies, just to decide whether or not to trace a request.
+
+=== OpenShift-based decision
+
+On this scenario, we delegate the decision to our runtime environment. Similar
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: "In this scenario"

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+to a simple A/B testing scenario, our deployment architecture would allow for
+quick changes in the routing decision. It does require a simple change to our
+application, but it's a small compromise for the benefits it provides.
+
+For that, we'll use one single Image Stream (`is`) with our code, two Deployment
+Configurations (`dc`), two Services (`svc`) and one Route object. Each one of our
+`dc` objects would have different values for an environment variable called
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: "have a different value for"

**Review by objectiser — APPROVED** [review]
Excellent blog post - only thing I would may be say is that potentially the no-tracer could simply not have the HAWKULAR_APM_... environment variables, and the test in the code could be based on whether the HAWKULAR_APM_URI env variable is present? Rather than adding a new boolean env var - but fine to publish as is (with a couple of minor changes).

**Comment by jpkrohling** [comment]
About checking `HAWKULAR_APM_URI`: for our own code, I'd agree with that, but for this post, I'm taking the position of an operations professional, so, having an env var with an explicit meaning ("turn off APM") is desirable here.

**Comment by objectiser** [comment]
Fair point - although then I would have called it `TRACER_ENABLED` :)

**Comment by pavolloffay** [comment]
Can you change  `weight: X` at runtime?

**Comment by pavolloffay** [comment]
I have a more general comment about this approach: 

Imagine there are 3 services A,B,C building one distributed app.  All are instrumented(i) and not(n) so there are Ai, An, Bi, Bn...

Request path should look like: `A->B->C`

Example request:  `An->Bi(trace is started)->Cn`

What I'm trying to say that sampling decision is not propagated. 

Is there a way in OC to group services into I(instrumented) N(not instrumented) so if the first request goes to a service from I then all following requests goes only to services from I?

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-03-17-teaser.png[Final result]
+endif::[]
+
+It's also a good practice to empower the admin, or an automated monitoring system,
+with the capability of adjusting the sampling based on the current conditions.
+For instance, we'd probably want to get more traces from a
+link:/blog/2017/02/13/monitoring-canary-deployments.html[newly deployed version]
+of an existing application. We might want to compare the performance and adjust
+the percentage in steps. Similarly, if we are experiencing traffic peaks and need
```
Comment (pavolloffay) [inline src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: If I get it correctly you are saying that samplig should be adjusted for performance reasons. I would mention that it should be adjusted to get a representative sample (of the population/all requests) constantly in time.

**Comment by objectiser** [comment]
Good point @pavolloffay - maybe the easiest approach is to use APMTracer in both trace and no-trace services, but the no-trace service has a Sampling.NEVER sampling policy? As this sampling policy only comes into play when the trace is started?

This also means only the initial service needs to have the pair of services, all other services in the chain are controlled based on the propagating sampling property.

**Comment by pavolloffay** [comment]
> Good point @pavolloffay - maybe the easiest approach is to use APMTracer in both trace and no-trace services, but the no-trace service has a Sampling.NEVER sampling policy? As this sampling policy only comes into play when the trace is started?

With this the result would be the same as in my previous comment. Ai->Bn sample header would be propagated to C which is good but it would produce "broken" trace. With this kind of setting it is hard to calculate overall percentage of all sampled traces. Actually not hard but you would get just lower number as expected.

**Comment by jpkrohling** [comment]
> Can you change weight: X at runtime?

Yes, either via `oc route edit`, or via the web UI.

> Is there a way in OC to group services into I(instrumented) N(not instrumented) so if the first request goes to a service from I then all following requests goes only to services from I?

Good question. I don't think there's a way in OpenShift to accomplish that without code changes within the services themselves (side by side with the business, or within the tracer). I would go a bit further than @objectiser : if this is a scenario we can see that people are interested on, we could have a special code on our tracers.

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc (hunk 5 above)
Comment (jpkrohling) [reply src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: Not only for performance reasons: performance is one of the reasons people might be interested in implementing a sampling strategy.

**Comment by pavolloffay** [comment]
> Yes, either via oc route edit, or via the web UI.

For me this is the crucial point of the blog :+1: Is it mentioned somewhere?

> Good question. I don't think there's a way in OpenShift to accomplish that without code changes within the services themselves (side by side with the business, or within the tracer). I would go a bit further than @objectiser : if this is a scenario we can see that people are interested on, we could have a special code on our tracers.

Then I think problem does not lie in tracer but you would have to make sure that all client requests resolves to right names (?) :)

**Comment by objectiser** [comment]
@pavolloffay It wouldn't be an issue, as there wouldn't be a Bn - only the A service would need `i` and `n` variants.

**Comment by jpkrohling** [comment]
@pavolloffay , I updated the blog post to include information about changing the route.

**Comment by jpkrohling** [comment]
@theute , @pilhuhn , would you please merge, if it looks good to you as well ?

**Comment by pavolloffay** [comment]
To sumarize: due to issue mentioned above sampled header should be propagated through services.

Therefore not instrumented services should be actually instrumented with sampling set to 0% and instrumented with sampling I gues 100%. This would allow you to controll sampling on load balancer. (Maybe there are also other solutions)

I apologize made a mistake in https://github.com/hawkular/hawkular.github.io/pull/286#issuecomment-287397889 spans from a service with sampling 0% would be reported as previous sampling decision takes precedence.

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc (hunk 5 above)
Comment (pilhuhn) [reply src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: Actually that may also work in the opposite direction - if you encounter an issue, you may want to trace every request to collect more evidence about the issue (e.g. to find out the issue only shows when ordering cars), while for normal operation you only sample every minute.

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+quick changes in the routing decision. It does require a simple change to our
+application, but it's a small compromise for the benefits it provides.
+
+For that, we'll use one single Image Stream (`is`) with our code, two Deployment
+Configurations (`dc`), two Services (`svc`) and one Route object. Each one of our
+`dc` objects would have a different value for an environment variable called
+`TRACER_ENABLED`. Each `svc` has its own `dc`, and the router sends a
+percentage of the requests to each of the `svc` instances. The main advantage of
+this approach is that the application image is the same for both `tracing-enabled`
+and `tracing-disabled` scenarios.
```
Comment (pilhuhn) [inline src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: Can you add an image with some boxes where you show the logical hierarchy that you describe?

Make explicit why the router is sending a percentage to each svc -- is that the 20% from above?

**Comment by pilhuhn** [comment]
Great! Thanks @jpkrohling

File: src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc (hunk 1 above)
Comment (josejulio) [reply src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc]: This part is still missing (create the repo or update the link)

**Pull request merged**